File: scripts/deploy.mjs
/**
 * @source <github.com/wikimedia-gadgets/xfdcloser/blob/master/bin/deploy.js>
 * @license MIT + CC-BY-SA-3.0 + CC-BY-4.0
 */
// This software is published under the following licenses. You may select the license of your choice.
// - Note: Files published on Wikipedia, including previous versions of XFDcloser, are also available under
// Creative Commons Attribution-ShareAlike 3.0 Unported License (CC BY-SA 3.0)
// https://creativecommons.org/licenses/by-sa/3.0/ and GNU Free Documentation License (GFDL)
// http://www.gnu.org/copyleft/fdl.html
// ---------------------------------------------------------------------------------------------------
// MIT License
// Copyright (c) 2020 Evad37
// Permission is hereby granted, free of charge, to any person obtaining a copy of this software
// and associated documentation files (the "Software"), to deal in the Software without
// restriction, including without limitation the rights to use, copy, modify, merge, publish,
// distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom
// the Software is furnished to do so, subject to the following conditions:
// The above copyright notice and this permission notice shall be included in all copies or
// substantial portions of the Software.
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
// BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
// ---------------------------------------------------------------------------------------------------
// Creative Commons Attribution 4.0 International Public License
// For a human-readable summary, see https://creativecommons.org/licenses/by/4.0/

/**
 * This script is used to deploy files to the wiki.
 * You must have interface-admin rights to deploy as gadget.
 *
 * ----------------------------------------------------------------------------
 *  Set up:
 * ----------------------------------------------------------------------------
 * 1) Use [[Special:BotPasswords]] to get credentials. Make sure you enable
 *  sufficient permissions.
 * 2) Create a JSON file to store the username and password. This should be
 *  a plain JSON object with keys "username" and "password", see README
 *  file for an example. Save it here in the "scripts" directory with file
 *  name "credentials.json".
 *
 * ---------------------------------------------------------------------------
 *  Pre-deployment checklist:
 * ---------------------------------------------------------------------------
 * 1) Changes committed and merged to master branch on GitHub repo
 * 2) Currently on master branch, and synced with GitHub repo
 * 3) Run a full build using "grunt build"
 * When all of the above are done ==> you are ready to proceed with deployment
 *
 * --------------------------------------------------------------------------
 *  Usage:
 * --------------------------------------------------------------------------
 * Ensure the pre-deployment steps above are completed, unless you are only
 * deploying to the testwiki (www.qiuwen.wiki). Then, run this script:
 * In the terminal, enter
 *   node deploy.js
 * and supply the requested details.
 * Notes:
 * - The default summary if not specified is "Updated from repository"
 * - Edit summaries will be prepended with the version number from
 *   the package.json file
 * - Changes to gadget definitions need to be done manually
 */
import { readFile as _readFile } from 'fs/promises';
import { mwn } from 'mwn';
import { execSync } from 'child_process';
import prompts from 'prompts';
import chalk from 'chalk';
import minimist from 'minimist';
import Credentials from './credentials.json' assert {type: 'json'};
import path from 'path';
const __dirname = path.resolve();

// Adjust target file names if necessary
// All file paths are with respect to repository root
const deployTargets = [{
	file: 'dist/twinkle.js',
	target: 'MediaWiki:Gadget-Twinkle.js'
}, {
	file: 'src/twinkle.css',
	target: 'MediaWiki:Gadget-Twinkle.css'
}, {
	file: 'dist/morebits.js',
	target: 'MediaWiki:Gadget-morebits.js'
}, {
	file: 'src/lib/select2.min.css',
	target: 'MediaWiki:Gadget-select2.min.css'
}, {
	file: 'dist/lib/select2.min.js',
	target: 'MediaWiki:Gadget-select2.min.js'
}, {
	file: 'src/morebits.css',
	target: 'MediaWiki:Gadget-morebits.css'
}, {
	file: 'src/twinkle-pagestyles.css',
	target: 'MediaWiki:Gadget-Twinkle-pagestyles.css'
}, {
	file: 'dist/modules/friendlytag.js',
	target: 'MediaWiki:Gadget-friendlytag.js'
}, {
	file: 'dist/modules/friendlytalkback.js',
	target: 'MediaWiki:Gadget-friendlytalkback.js'
}, {
	file: 'dist/modules/twinklearv.js',
	target: 'MediaWiki:Gadget-twinklearv.js'
}, {
	file: 'dist/modules/twinklebatchprotect.js',
	target: 'MediaWiki:Gadget-twinklebatchprotect.js'
}, {
	file: 'dist/modules/twinklebatchdelete.js',
	target: 'MediaWiki:Gadget-twinklebatchdelete.js'
}, {
	file: 'dist/modules/twinklebatchundelete.js',
	target: 'MediaWiki:Gadget-twinklebatchundelete.js'
}, {
	file: 'dist/modules/twinkleblock.js',
	target: 'MediaWiki:Gadget-twinkleblock.js'
}, {
	file: 'dist/modules/twinkleclose.js',
	target: 'MediaWiki:Gadget-twinkleclose.js'
}, {
	file: 'dist/modules/twinkleconfig.js',
	target: 'MediaWiki:Gadget-twinkleconfig.js'
}, {
	file: 'dist/modules/twinklecopyvio.js',
	target: 'MediaWiki:Gadget-twinklecopyvio.js'
}, {
	file: 'dist/modules/twinklediff.js',
	target: 'MediaWiki:Gadget-twinklediff.js'
}, {
	file: 'dist/modules/twinklefluff.js',
	target: 'MediaWiki:Gadget-twinklefluff.js'
}, {
	file: 'dist/modules/twinkleimage.js',
	target: 'MediaWiki:Gadget-twinkleimage.js'
}, {
	file: 'dist/modules/twinkleprotect.js',
	target: 'MediaWiki:Gadget-twinkleprotect.js'
}, {
	file: 'dist/modules/twinklespeedy.js',
	target: 'MediaWiki:Gadget-twinklespeedy.js'
}, {
	file: 'dist/modules/twinklestub.js',
	target: 'MediaWiki:Gadget-twinklestub.js'
}, {
	file: 'dist/modules/twinkleunlink.js',
	target: 'MediaWiki:Gadget-twinkleunlink.js'
}, {
	file: 'dist/modules/twinklewarn.js',
	target: 'MediaWiki:Gadget-twinklewarn.js'
}, {
	file: 'dist/modules/twinklexfd.js',
	target: 'MediaWiki:Gadget-twinklexfd.js'
}];

class Deploy {
	async deploy() {
		if (!isGitWorkDirClean()) {
			log('red', '[WARN] Git working directory is not clean.');
		}
		const config = this.loadConfig();
		await this.getApi(config);
		await this.login();
		await this.makeEditSummary();
		await this.savePages();
	}

	loadConfig() {
		try {
			return Credentials;
		} catch (e) {
			log('red', 'No credentials.json file found.');
			return {};
		}
	}

	async getApi(config) {
		this.api = new mwn(config);
		try {
			this.api.initOAuth();
			this.usingOAuth = true;
		} catch (e) {
			if (!config.username) {
				config.username = await input('> Enter username');
			}
			if (!config.password) {
				config.password = await input('> Enter bot password', 'password');
			}
		}
		if (args.testwiki) {
			config.apiUrl = `https://www.qiuwen.wiki/api.php`;
		} else {
			if (!config.apiUrl) {
				if (Object.keys(config).length) {
					log('yellow', 'Tip: you can avoid this prompt by setting the apiUrl as well in credentials.json');
				}
				const site = await input('> Enter sitename (eg. www.qiuwen.wiki)');
				config.apiUrl = `https://${site}/api.php`;
			}
		}
		this.api.setOptions(config);
	}

	async login() {
		this.siteName = this.api.options.apiUrl.replace(/^https:\/\//, '').replace(/\/.*/, '');
		log('yellow', '--- Logging in ...');
		if (this.usingOAuth) {
			await this.api.getTokensAndSiteInfo();
		} else {
			await this.api.login();
		}
	}

	async makeEditSummary() {
		this.sha = execSync('git rev-parse --short HEAD').toString('utf8').trim();
		const message = await input('> Edit summary message (optional): ');
		this.editSummary = `Git 版本 ${this.sha}: ${message || '代码仓库同步更新'}`;
		console.log(`Edit summary is: "${this.editSummary}"`);
	}

	async readFile(filepath) {
		return (await _readFile(__dirname + '/' + filepath)).toString();
	}

	async savePages() {
		await input(`> Press [Enter] to start deploying to ${this.siteName} or [ctrl + C] to cancel`);

		log('yellow', '--- starting deployment ---');

		for await (let { file, target } of deployTargets) {
			let fileText = await this.readFile(file);
			let fileHeader = '/* <nowiki> */\n';
			let fileFooter = '\n/* </nowiki> */';
			fileText = fileHeader + fileText + fileFooter;
			try {
				const response = await this.api.save(target, fileText, this.editSummary);
				if (response && response.nochange) {
					log('yellow', `━ No change saving ${file} to ${target} on ${this.siteName}`);
				} else {
					log('green', `✔ Successfully saved ${file} to ${target} on ${this.siteName}`);
				}
			} catch (error) {
				log('red', `✘ Failed to save ${file} to ${target} on ${this.siteName}`);
				logError(error);
			}
		}
		log('yellow', '--- end of deployment ---');
	}
}

const isGitWorkDirClean = () => {
	try {
		execSync('git diff-index --quiet HEAD --');
		return true;
	} catch (e) {
		return false;
	}
};

const input = async (message, type = 'text', initial = '') => {
	let name = String(Math.random());
	return (
		await prompts({
			type,
			name,
			message,
			initial,
		})
	)[name];
};

const logError = (error) => {
	error = error || {};
	console.log((error.info || 'Unknown error') + '\n', error.response || error);
};

const log = (color, ...args) => {
	console.log(chalk[color](...args));
};

const args = minimist(process.argv.slice(2));
new Deploy().deploy();

File: scripts/deploy.py
"""
本同步脚本文件采用GPLv3授权
@"license": <https://www.gnu.org/licenses/gpl-3.0.zh-cn.html>
@author: <github@Blossomstripe>
@source: <https://github.com/TopRealm/InterfaceCodes/blob/master/scripts/sync_mediawiki.py>
"""
import os
import subprocess as sp

import mwclient

# 常量部分
# 定义需在代码中额外插入的字符串
FILE_HEADER = "/* <nowiki> */\n"
FILE_FOOTER = "\n/* </nowiki> */"
FILE_WARNING = "/**\n * +--------------------------------------------------------+\n * |         === WARNING: GLOBAL GADGET FILE ===            |\n * +--------------------------------------------------------+\n * |      All changes should be made in the repository,     |\n * |              otherwise they will be lost.              |\n * +--------------------------------------------------------+\n * |      Changes to this page may affect many users.       |\n * |  Please discuss changes at talk page before editing.   |\n * +--------------------------------------------------------+\n */\n"
# 定义部署源文件、目标文件和授权协议
DEPLOY_TARGETS = [
    # twinkle
    {
        "file": 'distTS/src/twinkle/twinkle.js',
        "target": 'MediaWiki:Gadget-Twinkle.js',
        "license": 'src/licenseHeader'
    },
    {
        "file": 'src/twinkle/twinkle.css',
        "target": 'MediaWiki:Gadget-Twinkle.css',
        "license": 'src/licenseHeader'
    },
    {
        "file": 'src/twinkle/twinkle-pagestyles.css',
        "target": 'MediaWiki:Gadget-Twinkle-pagestyles.css',
        "license": 'src/licenseHeader'
    },
    # modules
    {
        "file": 'distTS/src/modules/friendlytag.js',
        "target": 'MediaWiki:Gadget-friendlytag.js',
        "license": 'src/licenseHeader'
    },
    {
        "file": 'distTS/src/modules/friendlytalkback.js',
        "target": 'MediaWiki:Gadget-friendlytalkback.js',
        "license": 'src/licenseHeader'
    },
    {
        "file": 'distTS/src/modules/twinklearv.js',
        "target": 'MediaWiki:Gadget-twinklearv.js',
        "license": 'src/licenseHeader'
    },
    {
        "file": 'distTS/src/modules/twinklebatchprotect.js',
        "target": 'MediaWiki:Gadget-twinklebatchprotect.js',
        "license": 'src/licenseHeader'
    },
    {
        "file": 'distTS/src/modules/twinklebatchdelete.js',
        "target": 'MediaWiki:Gadget-twinklebatchdelete.js',
        "license": 'src/licenseHeader'
    },
    {
        "file": 'distTS/src/modules/twinklebatchundelete.js',
        "target": 'MediaWiki:Gadget-twinklebatchundelete.js',
        "license": 'src/licenseHeader'
    },
    {
        "file": 'distTS/src/modules/twinkleblock.js',
        "target": 'MediaWiki:Gadget-twinkleblock.js',
        "license": 'src/licenseHeader'
    },
    {
        "file": 'distTS/src/modules/twinkleclose.js',
        "target": 'MediaWiki:Gadget-twinkleclose.js',
        "license": 'src/licenseHeader'
    },
    {
        "file": 'distTS/src/modules/twinkleconfig.js',
        "target": 'MediaWiki:Gadget-twinkleconfig.js',
        "license": 'src/licenseHeader'
    },
    {
        "file": 'distTS/src/modules/twinklecopyvio.js',
        "target": 'MediaWiki:Gadget-twinklecopyvio.js',
        "license": 'src/licenseHeader'
    },
    {
        "file": 'distTS/src/modules/twinklediff.js',
        "target": 'MediaWiki:Gadget-twinklediff.js',
        "license": 'src/licenseHeader'
    },
    {
        "file": 'distTS/src/modules/twinklefluff.js',
        "target": 'MediaWiki:Gadget-twinklefluff.js',
        "license": 'src/licenseHeader'
    },
    {
        "file": 'distTS/src/modules/twinkleimage.js',
        "target": 'MediaWiki:Gadget-twinkleimage.js',
        "license": 'src/licenseHeader'
    },
    {
        "file": 'distTS/src/modules/twinkleprotect.js',
        "target": 'MediaWiki:Gadget-twinkleprotect.js',
        "license": 'src/licenseHeader'
    },
    {
        "file": 'distTS/src/modules/twinklespeedy.js',
        "target": 'MediaWiki:Gadget-twinklespeedy.js',
        "license": 'src/licenseHeader'
    },
    {
        "file": 'distTS/src/modules/twinklestub.js',
        "target": 'MediaWiki:Gadget-twinklestub.js',
        "license": 'src/licenseHeader'
    },
    {
        "file": 'distTS/src/modules/twinkleunlink.js',
        "target": 'MediaWiki:Gadget-twinkleunlink.js',
        "license": 'src/licenseHeader'
    },
    {
        "file": 'distTS/src/modules/twinklewarn.js',
        "target": 'MediaWiki:Gadget-twinklewarn.js',
        "license": 'src/licenseHeader'
    },
    {
        "file": 'distTS/src/modules/twinklexfd.js',
        "target": 'MediaWiki:Gadget-twinklexfd.js',
        "license": 'src/licenseHeader'
    },
    # Morebits
    {
        "file": 'distTS/src/morebits/morebits.js',
        "target": 'MediaWiki:Gadget-morebits.js',
        "license": 'src/licenseHeader'
    },
    {
        "file": 'src/morebits/morebits.css',
        "target": 'MediaWiki:Gadget-morebits.css',
        "license": 'src/licenseHeader'
    },
    # Select2
    {
        "file": 'src/select2/select2.min.css',
        "target": 'MediaWiki:Gadget-select2.min.css',
        "license": 'src/select2/select2-licenseHeader'
    },
    {
        "file": 'src/select2/select2.min.js',
        "target": 'MediaWiki:Gadget-select2.min.js',
        "license": 'src/select2/select2-licenseHeader'
    }
]


def last_commit_info(file_name: str) -> tuple:
    """调用bash的git命令，获取给定文件的最后一次commit哈希值及留言。

    Args:
        file_name (str): 文件路径

    Raises:
        Exception: 若git返回的哈希值未通过格式检查，则抛出异常

    Returns:
        tuple: (获取的完整SHA-1哈希值, 留言)
    """
    # 若传入的文件位于distTS中，则定位其编译前的原始路径
    if file_name[:7] == "distTS/":
        file_name = file_name[7:]

    cmd = f"git log --pretty=format:\"%H %s\" -1 -- {file_name}"
    ans = sp.run(cmd, stdout=sp.PIPE, shell=True).stdout.decode("utf-8").strip()
    commit_hash = ans.split(" ", 1)[0]
    commit_msg = ans.split(" ", 1)[1]
    # 格式检查
    if commit_hash.isalnum() and len(commit_hash) == 40:
        return (commit_hash, commit_msg)
    else:
        raise Exception("获取Commit Hash失败：格式检查未通过")


def sync_file(site: mwclient.Site, page_name: str, text_new: str, file_name=None) -> None:
    """将传入的新内容与MediaWiki对应页面内容进行比对，若不一致则对页面进行更新。

    Args:
        site (mwclient.Site): 已登录的mwclient.Site对象
        page_name (str): 目标MediaWiki页面名称
        text_new (str): 用于比对的新内容
        file_name (str, optional): 文件名，用于调用last_commit_hash函数获取commit记录。默认值为None，即不在编辑摘要中记录commit哈希
    """
    page = site.pages[page_name]
    text_old = page.text()
    # MediaWiki会自动删除上传文本末尾的空白字符
    if text_old != text_new.rstrip():
        summary = "Git更新："
        if file_name is not None:
            commit_hash, commit_msg = last_commit_info(file_name)
            summary += f"([https://github.com/TopRealm/Twinkle/commit/{commit_hash} {commit_hash[:7]}]) {commit_msg}"
        page.edit(text_new, summary)
        print(page_name, "\t", "changed")
    else:
        print(page_name, "\t", "not_changed")


# 登录MediaWiki
site = mwclient.Site("youshou.wiki", path="/")
site.login("Github-bot", os.environ["MW_BOT_PASSWORD"])


# 同步仓库中的代码文件
for deploy_item in DEPLOY_TARGETS:
    with open(deploy_item["file"], "r", encoding="utf-8") as pfile:
        page_text = pfile.read()

    with open(deploy_item["license"], "r", encoding="utf-8") as pfile:
        license_text = pfile.read()
        page_text = license_text + FILE_WARNING + FILE_HEADER + page_text + FILE_FOOTER

    sync_file(site, deploy_item["target"], page_text, deploy_item["file"])
